Fix/transactions fetch

Lyncee59 opened the pull request:
## Description
- Fix - Problems related to the endless scroll on each transactions pages.

## Change Type
Please enter one or more of the following: 
- Bug Fix
- Upgrades

## Testing Steps
Detail the steps required for the reviewer(s) to verify and test these changes.

## Code Checklist
- [x] Code compiles successfully (verified via `yarn start`)
- [x] No lint issues exist (verified via `yarn lint`)
- [x] New and existing unit tests pass (verified via `yarn test`)
- [x] `README.md` and other documentation is updated as needed

plondon commented on packages/blockchain-wallet-v4/src/redux/data/bch/selectors.js:
```diff
@@ -17,17 +17,15 @@ export const getTransactionHistory = path([dataPath, 'bch', 'transaction_history
 export const getTotalTxPerAccount = curry((xpubOrAddress, state) => getAddresses(state).map(path([xpubOrAddress, 'n_tx'])))
 
+export const getFinalBalance = curry((address, state) => getAddresses(state).map(path([address, 'final_balance'])).map(x => x || 0))
+
```
Nice

plondon commented:
Noticed a problem when creating a wallet:
<img width="427" alt="screen shot 2018-06-18 at 9 03 19 pm" src="https://user-images.githubusercontent.com/5640772/41570015-19f9395c-733b-11e8-8a61-e7ec0d1c5102.png">

plondon commented:
BCH balance for unspendable and spendable is the same:
<img width="333" alt="screen shot 2018-06-18 at 9 08 26 pm" src="https://user-images.githubusercontent.com/5640772/41570209-ce567ab8-733b-11e8-98d8-0a8bae08250b.png">

plondon commented:
BCH txs scroll fails:
<img width="383" alt="screen shot 2018-06-18 at 9 12 38 pm" src="https://user-images.githubusercontent.com/5640772/41570342-61084be8-733c-11e8-8ac4-c0a14e95397b.png">

plondon commented:
Watch only transactions no longer show up in the recent activity list.

<img width="567" alt="screen shot 2018-06-18 at 9 16 44 pm" src="https://user-images.githubusercontent.com/5640772/41570461-f5042678-733c-11e8-891d-12a88d1fbafa.png">

Lyncee59 commented:
@plondon I addressed and fixed the following bugs:
- BCH scroll issue
- Wallet creation bug (indexOf): This bug was related under the hood to a sneaky bug in the metadataBCH creation happening when we register a wallet. As now the balance is relying directly on the BCH metadata to properly filter archived accounts, we could detect this bug that was not properly populating the BCH metadata after the initial creation of BCH metadata during wallet creation.

Could not reproduce/not longer an issue:
- Balance: could not reproduce, on my side balances including unspendable match development version
- Watchonly in activityList: could not reproduce, I can see watchonly addresses and the tooltip

plondon commented:
The top is this branch, the bottom is staging:

![screen shot 2018-06-20 at 10 01 51 am](https://user-images.githubusercontent.com/5640772/41663277-072c7e24-7471-11e8-903a-6806c3cb6a69.png)

plondon commented:
It's okay if BTC doesn't show if it's 0, think that's better actually. But BCH unspendable balance is incorrect.

plondon commented:
Left is staging, right is this branch.
It's fine to not show btc/bch unspendable if it's 0, probably better, but bch unspendable is wrong:
![screen shot 2018-06-20 at 10 08 30 am](https://user-images.githubusercontent.com/5640772/41663675-fc18c000-7471-11e8-862a-499d7a0a0700.png)

plondon commented on packages/blockchain-wallet-v4-frontend/src/layouts/Wallet/MenuTop/Balance/index.js:
```diff
@@ -1,25 +1,17 @@
   render () {
-    const { data } = this.props
-    return <Template btcUnspendableContext={data.btcUnspendableContext} bchUnspendableContext={data.bchUnspendableContext} path={data.path} />
+    return <Template path={this.props.path} />
```
Something is wrong here:
![screen shot 2018-06-20 at 10 13 12 am](https://user-images.githubusercontent.com/5640772/41663930-917e9a8e-7472-11e8-97b6-7e325841bec8.png)

Lyncee59 replied:
As explained earlier, there was a bug in the way we calculate the getSpendableContext and getUnspendableContext for BCH, and it was automatically excluding the legacyAddresses from the wallet payload. To double check carefully then, but these legacyAddresses were previously not included in the Spendable/Unspendable context (if not archived) and were influencing the final balance.

Lyncee59 replied:
Ok, I misunderstood the concept of Unspent balances and thought it was only about 'Imported addresses'. check my last changes

plondon commented:
Seems really close and fixes the original issue (1031) as well! Just a few comments

plondon commented on packages/blockchain-wallet-v4/package.json:
```diff
@@ -79,6 +79,7 @@
     "redux": "3.6.0",
     "redux-immutable": "3.1.0",
     "redux-saga": "0.16.0",
+    "reselect": "3.0.1",
```
Is this necessary in v4 core?

plondon commented on packages/blockchain-wallet-v4/src/redux/wallet/selectors.js:
```diff
@@ -27,6 +27,8 @@ export const getActiveAddresses = compose(ImtoJS, map(Address.toJS), AddressMap.
 export const getSpendableAddresses = compose(ImtoJS, map(Address.toJS), AddressMap.selectSpendable, Wallet.selectAddresses, getWallet)
 export const getUnspendableAddresses = compose(ImtoJS, map(Address.toJS), AddressMap.selectUnspendable, Wallet.selectAddresses, getWallet)
+export const getActiveSpendableAddresses = compose(ImtoJS, map(Address.toJS), AddressMap.selectActiveSpendable, Wallet.selectAddresses, getWallet)
+export const getActiveUnspendableAddresses = compose(ImtoJS, map(Address.toJS), AddressMap.selectActiveUnspendable, Wallet.selectAddresses, getWallet)
```
If an address is not active it is not spendbale

plondon commented:
I also don't see why we need
```
export const getActiveSpendableAddresses = compose(ImtoJS, map(Address.toJS), AddressMap.selectActiveSpendable, Wallet.selectAddresses, getWallet)
```

if we have 

```export const getSpendableContext = compose(ImtoJS, Wallet.selectSpendableContext, getWallet)```
